### PR TITLE
`enum Rav1dRestorationType`: backport optimization from dav1d 1.3.0

### PR DESCRIPTION
**Pull request by fbossen**
Embed `sgr_idx` into `enum Rav1dRestorationType` to avoid storing it separately. Define `enum Rav1dRestorationType` as `repr(u8)` to match C code and save some memory.

### CODE REVIEW

File: include/dav1d/headers.rs [hunk 1]
```diff
@@ -193,13 +193,38 @@ pub const DAV1D_RESTORATION_WIENER: Dav1dRestorationType =
 pub const DAV1D_RESTORATION_SGRPROJ: Dav1dRestorationType =
     Rav1dRestorationType::SgrProj as Dav1dRestorationType;
 
-#[derive(Debug, Clone, Copy, PartialEq, Eq, FromRepr, Default)]
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, FromRepr, Default)]
+#[repr(u8)]
```
Comment (kkysen) [inline include/dav1d/headers.rs:197]: We shouldn't need this.  It'll choose the smallest size anyways and specifying a repr disables niche optimizations I believe.

File: include/dav1d/headers.rs [hunk 2]
```diff
@@ -193,13 +193,38 @@ pub const DAV1D_RESTORATION_WIENER: Dav1dRestorationType =
...
+    SgrProj14,
+    SgrProj15,
+}
+
+impl Rav1dRestorationType {
+    pub fn discriminant(&self) -> u8 {
+        // SAFETY: Because `Self` is marked `repr(u8)`, its layout is a `repr(C)` `union`
+        // between `repr(C)` structs, each of which has the `u8` discriminant as its first
+        // field, so we can read the discriminant without offsetting the pointer.
+        unsafe { *<*const _>::from(self).cast::<u8>() }
```
Comment (kkysen) [inline include/dav1d/headers.rs:226]: We shouldn't use unsafe here.  You can just do `as u8` since the enum is data-less (only the variants).

**Review by kkysen — CHANGES_REQUESTED** [review]

**Review by kkysen — COMMENTED** [review]
@fbossen, could I push a change?  I think it's easier to show that way than explaining it in scattered comments.  I want to add an `enum SgrIdx` to simplify this.

**Comment by fbossen** [comment]
> @fbossen, could I push a change? I think it's easier to show that way than explaining it in scattered comments. I want to add an `enum SgrIdx` to simplify this.

No problem. Please do.

**Comment by fbossen** [comment]
> @fbossen, how does [f72d0b9](https://github.com/memorysafety/rav1d/commit/f72d0b902c4189202901f85f2fe47a4be1a04c2b) look?

LGTM

**Review by kkysen — APPROVED** [review]

**Pull request merged**